src/libpmemfile-core/block_array.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2016-2017, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "layout.h"
+#include "inode.h"
+#include "internal.h"
+#include "out.h"
+
+static void
+update_binfo(struct pmemfile_vinode *vinode)
+{
+	struct pmemfile_inode *inode = D_RW(vinode->inode);
+	struct block_info *binfo = &vinode->first_free_block;
+
+	if (binfo->arr != NULL) {
+		/*
+		 * If the block_info is not null, it means it was kept
+		 * up-to-date by the allocating routines.
+		 */
+		return;
+	}
+
+	/*
+	 * If binfo was not used before, it must be initialized.
+	 */
+
+	/*
+	 * Find the block_array containing the next free block metadata
+	 * slot. This is either the block_array stored right in the
+	 * inode, ...
+	 */
+	binfo->arr = &inode->file_data.blocks;
+	/*
+	 * ... or if there is more than one block_array, it is
+	 * the one linked to it with the next field.
+	 */
+	if (!TOID_IS_NULL(binfo->arr->next))
+		binfo->arr = D_RW(binfo->arr->next);
+
+	binfo->idx = 0;
+
+	/* Find the first free entry in the block array. */
+	while (binfo->idx < binfo->arr->length &&
+	    binfo->arr->blocks[binfo->idx].size != 0)
+		binfo->idx++;
+}
+
+static bool
+has_free_block_entry(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	return binfo->idx < binfo->arr->length;
+}
```

Comment:
After previous acquire_new_entry binfo->idx may point to used entry. Previous code handled it correctly.